cmat128/doc.go

```diff
@@ -0,0 +1,98 @@
+// Generated by running
+//  go generate github.com/gonum/matrix
+// DO NOT EDIT.
+//
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+//
+// Package cmat128 provides implementations of complex128 matrix structures and
+// linear algebra operations on them.
+//
+// Overview
+//
+// This section provides a quick overview of the cmat128 package. The following
+// sections provide more in depth commentary.
+//
+// cmat128 provides:
+//  - Interfaces for a complex Matrix
+//
+// BLAS and LAPACK
+//
+// BLAS and LAPACK are the standard APIs for linear algebra routines. Many
+// operations in cmat128 are implemented using calls to the wrapper functions
+// in gonum/blas/blas64 and gonum/lapack/lapack64. By default, blas64 and
+// lapack64 call the native Go implementations of the routines. Alternatively,
+// it is possible to use C-based implementations of the APIs through the respective
+// cgo packages and "Use" functions. The Go implementation of LAPACK makes calls
+// through blas64, so if a cgo BLAS implementation is registered, the lapack64
+// calls will be partially executed in Go and partially executed in C.
+//
+// Type Switching
+//
+// The Matrix abstraction enables efficiency as well as interoperability. Go's
+// type reflection capabilities are used to choose the most efficient routine
+// given the specific concrete types. For example, in
+//  c.Mul(a, b)
+// if a and b both implement RawMatrixer, that is, they can be represented as a
+// blas64.General, blas64.Gemm (general matrix multiplication) is called, while
+// instead if b is a RawSymmetricer blas64.Symm is used (general-symmetric
+// multiplication), and if b is a *Vector blas64.Gemv is used.
+//
+// There are many possible type combinations and special cases. No specific guarantees
+// are made about the performance of any method, and in particular, note that an
+// abstract matrix type may be copied into a concrete type of the corresponding
+// value. If there are specific special cases that are needed, please submit a
+// pull-request or file an issue.
+//
+// Invariants
+//
+// Matrix input arguments to functions are never modified. If an operation would
+// change Matrix data, that matrix will be the receiver of a function.
+//
+// For convenience, a matrix may be used as both a receiver and as an input, e.g.
+//  a.Pow(a, 6)
+//  v.SolveVec(a.T(), v)
+// though in many cases this will cause an allocation (see Element Aliasing).
+// An exception to this rule is Copy, which does not allow a.Copy(a.T()).
```

Comment:
Maybe we should add in here something like:
"Note that a consequence of these invariants is that not all calls may share receiver and input, as many operations have a resulting shape that is not equal to either of the input arguments. In particular, a.Mul(a, b) will panic if b is not square (as the resulting matrix does not have the size of a), and a.Stack(a, b) will always panic (since the resulting shape is longer than a)"